09-spring-boot-ehcache-xml/src/main/java/com/example/springbootehcache/ext/ehcache/EhcachejCacheSpringConfiguration.java

```diff
@@ -0,0 +1,117 @@
+package com.example.springbootehcache.ext.ehcache;
+
+import org.ehcache.core.spi.service.InstantiatorService;
+import org.ehcache.impl.internal.classes.DefaultInstantiatorService;
+import org.ehcache.spi.service.Service;
+import org.ehcache.spi.service.ServiceProvider;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.BeanClassLoaderAware;
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.beans.factory.BeanFactoryAware;
+import org.springframework.beans.factory.NoSuchBeanDefinitionException;
+import org.springframework.beans.factory.ObjectProvider;
+import org.springframework.boot.autoconfigure.AutoConfiguration;
+import org.springframework.boot.autoconfigure.cache.CacheProperties;
+import org.springframework.boot.autoconfigure.cache.JCacheManagerCustomizer;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.core.io.Resource;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+
+import javax.cache.CacheManager;
+import javax.cache.Caching;
+import javax.cache.configuration.MutableConfiguration;
+import javax.cache.spi.CachingProvider;
+import java.io.IOException;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Properties;
+
+@AutoConfiguration
+@ConditionalOnClass({org.ehcache.CacheManager.class})
+@ConditionalOnMissingBean(value = org.springframework.cache.CacheManager.class, name = "cacheResolver")
+@EnableConfigurationProperties(CacheProperties.class)
+@ConditionalOnProperty("spring.cache.jcache.config")
+public class EhcachejCacheSpringConfiguration implements BeanClassLoaderAware, BeanFactoryAware {
+  private ClassLoader beanClassLoader;
+  private BeanFactory beanFactory;
+
+  @Override
+  public void setBeanClassLoader(ClassLoader classLoader) {
+    this.beanClassLoader = classLoader;
+  }
+
+  @Override
+  public void setBeanFactory(BeanFactory beanFactory) throws BeansException {
+    this.beanFactory = beanFactory;
+  }
+
+  @Bean
+  @ConditionalOnMissingBean
+  public InstantiatorService instantiatorService() {
+    return new DefaultInstantiatorService();
```

Comment:
It's not clear to me what the right sequence of delegation is here. Right now we're saying:

> BeanFactory first, then, if available, the users `InstantiatorService` bean otherwise the `DefaultInstantiatorService`.

Would this be better?

> The users InstantiatorService bean (if available), otherwise the `beanFactory` delegating to the `DefaultInstantiatorService`

Also, what about if the user has tried to override the instantiator service using the Ehcache dependency mechanism? Should it then be:

> The users InstantiatorService bean (if available), otherwise the `beanFactory` delegating to the service found using the Ehcache system?